Add a cast-to-int for extra safety before forming new DateTime object.

barryhughes opened the pull request:
Adds a small piece of defensive coding for the `action_scheduler_retention_period` hook. 

We use this hook to allow an integer value (representing a number of seconds) to be modified by other plugins, but it is possible they will return a non-integer and, in some cases—such as if `null` or `(string) ''` are returned—they will trigger a fatal error.

Closes https://github.com/woocommerce/action-scheduler/issues/917.

### Testing

Try this multi-liner:

```sh
wp eval "\
  add_filter( 'action_scheduler_retention_period', '__return_false' ); \
  ( new ActionScheduler_QueueRunner )->run(); \
"
```

With this branch, there should be no problem. Without this branch, you should find a fatal error similar to the following is reported:

```
Error: There has been a critical error on this website.Learn more about troubleshooting WordPress.Error details:Array
(
    [type] => 1
    [message] => Uncaught Exception: DateTime::__construct(): Failed to parse time string ( seconds ago) at position 0 (s)
    ...
)
```

### Changelog

> Fix - Improves safety around the `action_scheduler_retention_period` filter hook.

barryhughes committed:
Add a cast-to-int for extra safety before forming new DateTime object.

coreymckrill reviewed (commented):
👍 Works via command line testing.

Worth adding a unit test for this? Maybe [here](https://github.com/woocommerce/action-scheduler/blob/8bf7ceba4bbb3af1b0987b6c184c6d89710e1d15/tests/phpunit/runner/ActionScheduler_QueueCleaner_Test.php).

barryhughes commented:
Hi @coreymckrill,

> Worth adding a unit test for this? Maybe [here](https://github.com/woocommerce/action-scheduler/blob/8bf7ceba4bbb3af1b0987b6c184c6d89710e1d15/tests/phpunit/runner/ActionScheduler_QueueCleaner_Test.php).

Yep, I didn't because, honestly, I was struggling to think of a graceful way of doing this:

```php
add_filter( 'action_scheduler_retention_period', '__return_null' );
$cleaner = new ActionScheduler_QueueCleaner( $store );

# This is where the failure would happen; but it returns void so we
# cannot currently test the return value:
$cleaner->delete_old_actions();

# Simply asserting for the sake of it seems inelegant and bad form,
# though maybe I'm overthinking it/maybe we don't *need* the assertion:
$this->assertTrue( true, 'A queue cleaner will not fail even if a non-integer value like null is supplied as the retention period.');

remove_filter( 'action_scheduler_retention_period', '__return_null' );
```

A possibility is we could change `$cleaner->delete_old_actions()` to return the result (an array) of underlying call to the `clean_actions()` method, and since in this case we don't actually need to change the method signature to do this it would be backwards-compatible: then we can at least assert that an array was returned.

What do you think?

coreymckrill commented:
@barryhughes 

> A possibility is we could change $cleaner->delete_old_actions() to return the result (an array) of underlying call to the clean_actions() method

It seems like it would be fine to do this, and would just generally make the method more testable.

One other possible approach might be to do a try/catch within the unit test:

```
add_filter( 'action_scheduler_retention_period', '__return_null' );
$cleaner = new ActionScheduler_QueueCleaner( $store );

try {
	$cleaner->delete_old_actions();
	$this->assertTrue( true );
} catch ( \Exception $e ) {
	$this->assertTrue( false );
}

remove_filter( 'action_scheduler_retention_period', '__return_null' );
```

Although maybe this is still a little bit hacky?

barryhughes committed:
Update ActionScheduler_QueueCleaner::delete_old_actions() to return an array of cleaned actions.

Additionally, add tests relating to safety around the cleanup retention period hook.

barryhughes commented:
Perhaps we've found a real world case for `doNotExpectException()` 🧐 ... I'll make a change and add _something_ to cover this.

barryhughes committed:
Guard against possible exceptions.

barryhughes commented:
Sorry, hitting pause: I think this can be improved some more.

barryhughes commented on classes/ActionScheduler_QueueCleaner.php:
```diff
@@ -40,16 +40,32 @@ public function __construct( ActionScheduler_Store $store = null, $batch_size =
+		try {
+			$cutoff = as_get_datetime_object( $lifespan . ' seconds ago' );
+		} catch ( Exception $e ) {
+			_doing_it_wrong(
```
I added this for completeness, in but didn't cover it in our unit tests because it seems tricky to reliably trigger it across different PHP runtimes (the behavior when casting to int from an unsupported type is [undefined](https://www.php.net/manual/en/language.types.integer.php#language.types.integer.casting)).

barryhughes committed:
Maintain the <= 3.5.4 level of strictness re retention periods, but avoid fatals.

barryhughes commented:
Apologies for the flood of changes 😬 ... to summarize:

- I removed the cast to int.
- This maintains the same level of strictness as previously (so, instead of suddenly allowing `null` as a retention period, which was previously disallowed, and assuming it should be treated as zero, we continue to disallow it).
- However, to prevent fatal errors, we now catch any exception that is raised when creating the `DateTime` and raise attention via `_doing_it_wrong`. 

This should maintain the same strictness, without the site breaking fatal.

coreymckrill reviewed (approved):
Looks great now! Doing the try/catch within the method instead of casting seems like a better approach.
